### PR TITLE
Implement the `!` type

### PR DESCRIPTION
**Pull request by canndrew**
This implements the never type (`!`) and hides it behind the feature gate `#[feature(never_type)]`. With the feature gate off, things should build as normal (although some error messages may be different). With the gate on, `!` is usable as a type and diverging type variables (ie. types that are unconstrained by anything in the code) will default to `!` instead of `()`.

(tracking issue https://github.com/rust-lang/rust/issues/35121 )

### CODE REVIEW

**Comment by canndrew** [comment]
How much effort is it worth going to to get this working without MIR? It works currently except for one weirdly failing test. But if we're going to be launching MIR soon is it worth keeping old trans working at all?

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -699,6 +699,39 @@ mod impls {
 
     ord_impl! { char usize u8 u16 u32 u64 isize i8 i16 i32 i64 }
 
+    // Note: This macro is a temporary hack that can be remove once we are building with a compiler
+    // that supports `!`
+    macro_rules! argh {
```
Comment (eddyb) [inline src/libcore/cmp.rs]: This can be a bit more formal, e.g. "Remove this in the next snapshot." and `not_stage0` instead of `args`.
cc @alexcrichton Does "SNAP" work anymore?

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -1363,6 +1363,29 @@ macro_rules! fmt_refs {
 
 fmt_refs! { Debug, Display, Octal, Binary, LowerHex, UpperHex, LowerExp, UpperExp }
 
+// Note: This macro is a temporary hack that can be remove once we are building with a compiler
+// that supports `!`
+macro_rules! argh {
```
Comment (eddyb) [inline src/libcore/fmt/mod.rs]: Same here.

File: src/libcore/lib.rs [hunk 3]
```diff
@@ -87,6 +87,7 @@
 #![feature(staged_api)]
 #![feature(unboxed_closures)]
 #![feature(question_mark)]
+#![cfg_attr(not(stage0), feature(bang_type))]
```
Comment (eddyb) [inline src/libcore/lib.rs]: Maybe a snapshot-related comment here.

File: src/librustc/ty/adjustment.rs [hunk 4]
```diff
@@ -106,6 +107,10 @@ pub struct AutoDerefRef<'tcx> {
 impl<'tcx> AutoAdjustment<'tcx> {
     pub fn is_identity(&self) -> bool {
         match *self {
+            AdjustEmptyToAny(ref ty) => match ty.sty {
```
Comment (eddyb) [inline src/librustc/ty/adjustment.rs]: You should never match against `AdjustEmptyToAny` with `ref`, `Ty` is a pointer itself.

File: src/librustc/ty/context.rs [hunk 5]
```diff
@@ -1245,6 +1247,10 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self.mk_ty(TySlice(ty))
     }
 
+    pub fn mk_empty(&self) -> Ty<'tcx> {
+        self.mk_ty(TyEmpty)
+    }
```
Comment (eddyb) [inline src/librustc/ty/context.rs]: Not sure we should have this if `tcx.types.empty` works.

File: src/librustc/ty/sty.rs [hunk 6]
```diff
@@ -927,11 +888,15 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
...
     }
 
-    pub fn is_empty(&self, _cx: TyCtxt) -> bool {
+    pub fn is_empty(&self, cx: TyCtxt) -> bool {
         // FIXME(#24885): be smarter here
         match self.sty {
             TyEnum(def, _) | TyStruct(def, _) => def.is_empty(),
-            _ => false
+            TyEmpty => true,
+            TyTuple(ref tys) => tys.iter().any(|ty| ty.is_empty(cx)),
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: My suggestion is to use `is_never` for `!` and leave `is_empty` untouched.

File: src/librustc_mir/build/expr/into.rs [hunk 7]
```diff
@@ -45,6 +45,26 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
+                };
+
+                unpack!(block = this.as_rvalue(block, source));
+
+                // This is an optimization. If the expression was a call then we already have an
+                // unreachable block. Don't bother to terminate it and create a new one.
+                if is_call {
+                    block.unit()
+                }
+                else {
```
Comment (eddyb) [inline src/librustc_mir/build/expr/into.rs]: Preferred style is `} else {`.

File: src/librustc_mir/build/mod.rs [hunk 8]
```diff
@@ -287,7 +287,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
     fn finish(self,
               upvar_decls: Vec<UpvarDecl>,
               arg_decls: IndexVec<Arg, ArgDecl<'tcx>>,
-              return_ty: ty::FnOutput<'tcx>)
+              return_ty: ty::Ty<'tcx>)
```
Comment (eddyb) [inline src/librustc_mir/build/mod.rs]: `Ty` should be written unprefixed (here and everywhere else in this PR).

File: src/librustc_mir/hair/cx/expr.rs [hunk 9]
```diff
@@ -60,6 +60,15 @@ impl<'tcx> Mirror<'tcx> for &'tcx hir::Expr {
                     kind: ExprKind::UnsafeFnPointer { source: expr.to_ref() },
                 };
             }
+            Some(&ty::adjustment::AdjustEmptyToAny(..)) => {
+                let adjusted_ty = cx.tcx.expr_ty_adjusted(self);
```
Comment (eddyb) [inline src/librustc_mir/hair/cx/expr.rs]: Should just get the type from the adjustment you already have.

File: src/librustc_typeck/check/mod.rs [hunk 10]
```diff
@@ -1522,6 +1517,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         self.tables.borrow_mut().node_types.insert(node_id, ty);
     }
 
+    #[inline]
+    pub fn write_ty_expr(&self, node_id: ast::NodeId, ty: Ty<'tcx>) {
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Do we need this method? Couldn't it just check if there's a `NodeExpr` in `tcx.map` for that ID?

File: src/librustc/ty/sty.rs (hunk 6 above)
Comment (canndrew) [inline src/librustc/ty/sty.rs]: > Why did you make this recursive?

To make it smarter, as the FIXME suggests. And because some tuples are empty.

> Also, you *shouldn't* use the same method for checking for `enum Void {}` and `!`. It would mean `Void` gets the same semantics as `!` which is not what you want.

The are some cases where we do want `enum Void {}` and `!` to have the same semantics. For example, a function that returns *any* empty type should be marked with the `NoReturn` attribute.

> My suggestion is to use `is_never` for `!` and leave `is_empty` untouched.

Well there's other places where I check for `!` specifically by matching on `sty`.

Comment (eddyb) [inline src/librustc/ty/sty.rs]: > To make it smarter, as the FIXME suggests.

That shouldn't be part of this PR, it introduces certain risks. The code is not prepared to deal with something like this. We need to wait until old trans is gone and then properly introduce enum optimizations around `!` and `Void`.

> Well there's other places where I check for `!` specifically by matching on `sty`.

It should be uniform, otherwise it gets confusing when it's okay to use which.

File: src/librustc_mir/hair/cx/expr.rs (hunk 9 above)
Comment (canndrew) [inline src/librustc_mir/hair/cx/expr.rs]: Ah, this and that other bug were because I took the type out of `AdjustEmptyToAny` and put it back in later.

File: src/librustc_typeck/check/mod.rs (hunk 10 above)
Comment (canndrew) [inline src/librustc_typeck/check/mod.rs]: Can we? I'll do that then.

Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Yeah, it should work the same, just without needing manual changes everywhere.

Comment (canndrew) [inline src/librustc_typeck/check/mod.rs]: @eddyb That doesn't seem to work. I changed `write_ty` to this:

```rust
#[inline]
pub fn write_ty(&self, node_id: ast::NodeId, ty: Ty<'tcx>) {
    debug!("write_ty({}, {:?}) in fcx {}",
           node_id, ty, self.tag());
    self.tables.borrow_mut().node_types.insert(node_id, ty);

    // Add adjustments to !-expressions
    if ty.is_never() {
        if let hir::map::NodeExpr(_) = self.tcx.map.find(node_id).unwrap() {
            self.write_adjustment(node_id, adjustment::AdjustNeverToAny(self.next_diverging_ty_var()));
        }
    }
}
```

And I'm getting "called `Option::unwrap()` on a `None` value". If `map.find` is returning `None` then I don't know whether to write the adjustment or not.

Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: `if let Some(hir::map::NodeExpr(_)) = self.tcx.map.find(node_id) {`

Comment (canndrew) [inline src/librustc_typeck/check/mod.rs]: Why is it okay to ignore `None` though? Can we be sure it's not an expression in that case?

Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Yes, if it were an expression, it would be in the map.

File: src/libcore/cmp.rs [hunk 11]
```diff
@@ -699,6 +699,39 @@ mod impls {
...
+            impl Ord for ! {
+                fn cmp(&self, _: &!) -> Ordering {
+                    *self
+                }
+            }
+        }
+    }
+
+    #[cfg(not(stage0))]
+    not_stage0!();
```
Comment (eddyb) [inline src/libcore/cmp.rs]: @nikomatsakis What do you think about these impls?

File: src/librustc/ty/sty.rs [hunk 12]
```diff
@@ -927,11 +888,27 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
...
-            _ => false
+            TyNever => true,
+
+            // FIXME(canndrew): There's no reason why this can't be uncommented, it's tested and it
+            // doesn't break anything. But I'm keeping my changes small for now.
+            //TyTuple(ref tys) => tys.iter().any(|ty| ty.is_uninhabited(cx)),
+
+            // FIXME(canndrew): this line breaks core::fmt
+            //TyRef(_, ref tm) => tm.ty.is_uninhabited(cx),
+            _ => false,
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: IMO this function should remain untouched, it *specifically* looks for `enum Foo {}` and *any* change in it could have unintended consequences that we are *not* prepared for.

File: src/librustc_const_eval/check_match.rs [hunk 13]
```diff
@@ -215,17 +215,17 @@ fn check_expr(cx: &mut MatchCheckCtxt, ex: &hir::Expr) {
             // Check for empty enum, because is_useful only works on inhabited types.
             let pat_ty = cx.tcx.node_id_to_type(scrut.id);
             if inlined_arms.is_empty() {
-                if !pat_ty.is_empty(cx.tcx) {
+                if !pat_ty.is_uninhabited(cx.tcx) {
                     // We know the type is inhabited, so this must be wrong
                     let mut err = struct_span_err!(cx.tcx.sess, ex.span, E0002,
-                                                   "non-exhaustive patterns: type {} is non-empty",
+                                                   "non-exhaustive patterns: type {} is inhabited",
```
Comment (eddyb) [inline src/librustc_const_eval/check_match.rs]: I don't thinks this PR should change the terminology here, yet.

File: src/libsyntax/print/pprust.rs [hunk 14]
```diff
@@ -976,6 +976,9 @@ impl<'a> State<'a> {
                 try!(self.print_opt_lifetime(lifetime));
                 try!(self.print_mt(mt));
             }
+            ast::TyKind::Never => {
+                word(&mut self.s, "!")?;
```
Comment (eddyb) [inline src/libsyntax/print/pprust.rs]: This should use `try!` like all the code around it.

File: src/libcore/cmp.rs (hunk 11 above)
Comment (canndrew) [inline src/libcore/cmp.rs]: I'm pretty sure I added these out of necessity to make a test pass. Also `!` is trivially totally ordered so it should impl `Ord`.

File: src/librustc/ty/sty.rs [hunk 15]
```diff
@@ -930,11 +933,15 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
-    pub fn is_empty(&self, _cx: TyCtxt) -> bool {
+    pub fn is_empty(&self, cx: TyCtxt) -> bool {
```
Comment (nikomatsakis) [inline src/librustc/ty/sty.rs]: I think that this function should be renamed -- it sounds very much like it is checking for `TyEmpty`, when in fact it does more (not sure if this happens in a later commit)

File: src/librustc_trans/debuginfo/metadata.rs [hunk 16]
```diff
@@ -931,6 +933,7 @@ fn basic_type_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
     debug!("basic_type_metadata: {:?}", t);
 
     let (name, encoding) = match t.sty {
+        ty::TyEmpty => ("!", DW_ATE_unsigned),
```
Comment (nikomatsakis) [inline src/librustc_trans/debuginfo/metadata.rs]: This debuginfo seems mildly dubious. "unsigned"? Is that really the best fit?

File: src/librustc_typeck/check/mod.rs [hunk 17]
```diff
@@ -1709,6 +1722,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn expr_ty(&self, ex: &hir::Expr) -> Ty<'tcx> {
+        if let Some(&adjustment::AdjustEmptyToAny(ref t))
+                = self.tables.borrow().adjustments.get(&ex.id) {
+            return t;
+        }
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: Huh. Kind of wacky that we only consider *this* kind of adjustment -- do we believe that this will be the only case in which an adjustment is present?

Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: I suggested this approach. The existing code might misbehave if `expr_ty` is sometimes adjusted.

File: src/libsyntax/visit.rs [hunk 18]
```diff
@@ -4,7 +4,8 @@
 //
 // Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
 // http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
-// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your,
+//                 "The `!` type is experimental");
```
Comment (nikomatsakis) [inline src/libsyntax/visit.rs]: is this fixed later?

File: src/librustc_typeck/check/mod.rs [hunk 19]
```diff
@@ -1709,6 +1717,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn expr_ty(&self, ex: &hir::Expr) -> Ty<'tcx> {
+        if let Some(&adjustment::AdjustNeverToAny(ref t))
+                = self.tables.borrow().adjustments.get(&ex.id) {
+            return t;
+        }
         match self.tables.borrow().node_types.get(&ex.id) {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: I am somewhat surprised that `node_ty` was not changed (and also surprised that `expr_ty` doesn't just call `node_ty`, I wonder when/why that changed).

@eddyb any thoughts on this point?

Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: I feel like `fcx.node_ty(expr.id)` and `fcx.expr_ty(expr)` are basically equivalent, and suspect there are bits of code that work over both patterns/expressions and just carry an `id` (and they would see the type as `!` and not the type variable). Though ... hmm .. at least `regionck` then goes and adjusts the type, which would be bad.

I wonder if we can rename `expr_ty` to `expr_ty_adjusted` and just have it apply adjustments?

**Comment by nikomatsakis** [comment]
OK so @eddyb and I chatted a bunch on IRC. In general, I'm feeling nervous about how the "never-to-any" adjustment is different from other adjustments. For one thing, it's applied eagerly, in `write_ty`, and we want it to be reflected in `expr_ty` (which until now was basically an alias for `node_ty`). This also seems to suggest that some of the other logic for applying adjustments may be wrong -- although it seems like we already in some cases apply coercions even when another coerce exists, and the code is handling that case.

I can see various ways to go forward:

- Just merge it. Typeck is messy, gets a bit messier, up the priority on some kind of cleanup rewriting. I feel nervous about this, but it'd probably be ok. 
- Rename `expr_ty` to `expr_ty_adjusted` and have it apply all adjustments. Rewrite uses of `expr_ty` to use that. This...would probably be more-or-less the same as now, because in general the adjustments haven't been applied (I think?) at the places where `expr_ty` is called (or, if they have, maybe we actually want to be seeing them?). 
- Rewrite to return a `Ty<'tcx>` out of the "check expr" routine and instead of calling `expr_ty` everywhere, actually use the (adjusted) return value. 

Honestly without trying to work through either of those latter two options, I'm not quite sure how easy/hard they would be, but it seems like option 3 probably leaves us with a cleaner setup than we started with overall (presuming it works), so maybe worth a try.

**Comment by canndrew** [comment]
> Honestly without trying to work through either of those latter two options, I'm not quite sure how easy/hard they would be, but it seems like option 3 probably leaves us with a cleaner setup than we started with overall (presuming it works), so maybe worth a try.

The way I started implementing this is: keep adding the `AdjustNeverToAny` coercion eagerly in `write_ty` and make `write_ty` return the coerced type; make the `check_*` functions return the type that they write; get rid of `expr_ty` and instead use the returned types everywhere. It might also be worth asserting that node types never get written twice and then going through and fixing the breakages that this causes. Is this all about right?

Would it be okay to merge this PR first and then I'll immediately go on with the refactoring? It just feels like this PR has grown in scope quite a bit, I don't entirely understand what I'm doing and there's a lot of opportunity to introduce new bugs. I'm worried that with the time and work it takes to do this the PR might get a lot bigger or get painfully out of sync with master.

**Comment by eddyb** [comment]
> keep adding the `AdjustNeverToAny` coercion eagerly in `write_ty` and make `write_ty` return the coerced type

The advantage of returning types is that you can pretty much get away with exactly one write for all expressions, one write for all patterns, etc., by choosing a choke-point (such as `check_expr_with_expectation_and_lvalue_pref`, after the `match`).
You then only need to have the *expression* `AdjustNeverToAny` logic once in the code, at the end of `check_expr_with_expectation_and_lvalue_pref`.

**Comment by canndrew** [comment]
@eddyb I think the next step after everything I just described would be to try and factor `write_ty` out of the giant match block in `check_expr`. It's non-trivial though because some of those writes happen inside other functions, some nodes get written multiple times etc.

**Comment by nikomatsakis** [comment]
@eddyb ok so as discussed on IRC, I am in favor of landing this RFC, on the assumption that we will address some of the thornier issues later. 

One thing I think is missing is negative tests. I'd like to see at least some tests that if you try to return incompatible types that will result in an error. Probably some (more?) tests targeting "Auto trait" inference -- and the limits, e.g. around specialization -- are warranted as well. 

One thing I have done in the past -- but didn't do yet for this PR -- is to make a list, based on the RFC, not the code, of things that might need to be tested, and then compare the code against that list. I'm happy to go through and do it, or you can and post it somewhere. :)

**Comment by nikomatsakis** [comment]
(In case it wasn't clear, the code itself seems good to me! Nice work as usual.)

**Comment by nikomatsakis** [comment]
Oops, I meant those comments for `impl Trait` :) but they mostly apply here as well. D'oh. Too many tabs.

**Comment by nikomatsakis** [comment]
OK, so, as discussed on IRC, I'm amenable to landing this code as-is and doing a follow-up to address the question of threading the expr type by returning values instead of calling `expr_ty`. We also need to address future-proof warnings.

In the meantime, I went through the RFC and tried to come up with a list of test cases that would be good to include. Some of them may already exist, so if you think so, just check them off.

- [ ] Positive and negative tests using `!` in various positions where it is not allowed today:
    - [ ] `let x: ! = panic!()`, as in the initial motivating example from the RFC
    - [ ] `let x: ! = 22`, should error out
    - [ ] `fn foo(x: !)`, this should be legal to declare and call with `panic!()` but illegal to call with 22`
    - [ ] `trait Foo { type T = !; fn bar() -> Self::T { ... } }`, similarly can have some pos/neg tests here
- [ ] `impl SomeTrait for !` -- use the example around `Option` from the RFC
- [ ] `impl SomeTrait<!> for T` -- similar to `PartialEq`, perhaps?
- [ ] Dead-code warnings (are these implemented?)
    - [ ] `let t = spawn(|| -> Result<u32, !> { Ok(22) }); t.unwrap_err(); println!("foo");` as in RFC
- [ ] Tests around the changing behavior of `return/break`
    - `match break { () => { } } // legal now, and afterwards I guess`
    - `match break { } // legal only with feature-gate enabled`
- [ ] Various small snippets from RFC:
    - [ ] `let x: u32 = break as u32;`
    - [ ] `match x { Ok(x) => ... Err(y) => ... }` where `x` has type `Result<T, !>`
        - [ ] Perhaps a test showing that different references to `y` have independent types
            - I think this is true, right? i.e., the type of `y` is `!`, and hence each ref to `y` gets a fresh type variable; this seems ok and like something we can commit to (same is true for fn args)
- [ ] Feature-gating tests:
    - [ ] show that old fallback still triggers w/o feature-gate
    - [ ] show error w/ feature-gate when expected (because of changed fallback)
    - [ ] show using `!` in various new possible locations errors w/o feature-gate

**Comment by canndrew** [comment]
Implementing these tests has brought to light a couple of issues. (0) Trait resolution doesn't seem to always work properly on `!`. This works: `<! as StaticMethod>::static_method()` but this doesn't `!::static_method()`. (1) Method calls that return `!` don't mark the rest of the block as unreachable, only function calls do. (2) `panic!() as T` ICEs the compiler.

None of these seem to be blocking issues though. The feature-gating seems to be working so would it make sense to merge this (with the new tests which pass) and I'll submit the bugfixes for the above issues along with their tests in separate PR(s)?

**Comment by canndrew** [comment]
Also, `match break {}` doesn't currently work because `Ty::is_uninhabited` returns false for `!`. I know both @nikomatsakis and @eddyb expressed reservations about extending this function to support anything other than empty structs and enums but I really don't think it's such a scary thing to do. In the version of this that got cratered `Ty::is_uninhabited` would also return true for both `!` and tuples where any of their members were uninhabited and it didn't cause any issues.

**Comment by petrochenkov** [comment]
@canndrew 
`!::static_method()` isn't supposed to work (compare with `()::default()`, it doesn't work either), non-path types need to be enclosed with angle brackets - `<!>::static_method`.

**Comment by canndrew** [comment]
@petrochenkov Oh, right. Well that's one bug fixed.

**Comment by canndrew** [comment]
Mkay, issue (0) wasn't an issue (thanks @petrochenkov), issue (1) should now be fixed, that just leaves issue (2). A testcase for this is that in this code the dead code will get detected:

```rust
let x: Result<u32, !> = Ok(23);
Result::<u32, !>::unwrap_err(x);
println!("dead code detected");
```

But in this it won't:

```rust
let x: Result<u32, !> = Ok(23);
Result::unwrap_err(x);
println!("dead code NOT detected");
```

Nor with this:

```rust
let x: Result<u32, !> = Ok(23);
x.unwrap_err();
println!("dead code NOT detected");
```

This happens with functions aswell. So this gets detected:

```rust
fn smash<T>(x: Result<u32, T>) -> T {
    x.unwrap_err()
}

let x: Result<u32, !> = Ok(23);
smash::<T>(x);
println!("dead code detected");
```

But not this:

```rust
fn smash<T>(x: Result<u32, T>) -> T {
    x.unwrap_err()
}

let x: Result<u32, !> = Ok(23);
smash(x);
println!("dead code NOT detected");
```

Any ideas? And should we consider the PR blocked on this?

**Comment by nikomatsakis** [comment]
> Any ideas? And should we consider the PR blocked on this?

Sorry, I didn't have time today (and probably won't) to read into the code and understand just why those examples work out the way they do -- I'm guessing it's because the dead-code detection is old and grody though. =) But in any case I don't see why they should block the PR. We can just add that to the tracking issue.

File: src/test/compile-fail/call-fn-never-arg-wrong-type.rs [hunk 20]
```diff
@@ -8,8 +8,13 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-fn forever2() -> ! { //~ ERROR computation may converge in a function marked as diverging
-  loop { break }
+#![feature(never_type)]
+
```
Comment (nikomatsakis) [inline src/test/compile-fail/call-fn-never-arg-wrong-type.rs]: These tests are great. My one nit: I'd prefer to see comments at the top explaining what precisely the test is trying to test. I find that so helpful as a reader...

File: src/test/compile-fail/never-fallback.rs [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(never_type)]
```
Comment (nikomatsakis) [inline src/test/compile-fail/never-fallback.rs]: What is this test testing? I don't see any panics or `!` in here..?

File: src/test/compile-fail/return-from-diverging.rs [hunk 22]
```diff
@@ -8,9 +8,10 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// pretty-expanded FIXME #23616
+fn fail() -> ! {
+    return "wow"; //~ ERROR mismatched types
```
Comment (nikomatsakis) [inline src/test/compile-fail/return-from-diverging.rs]: Hmm. I wonder if we want a customized error message in this scenario -- seems likely (and probably applies to all uses of `!` equally). Maybe another good item to follow for follow-up.

cc @jonathandturner -- maybe we can add such a thing to some list :)

File: src/test/compile-fail/never-fallback.rs (hunk 21 above)
Comment (canndrew) [inline src/test/compile-fail/never-fallback.rs]: It's testing that the diverging type of `_` on line 30 defaults to `!` and not `()`. The file will build without `#![feaure(never_type)]` - it's the same as the run-pass/unit-fallback.rs test that I added except for that.

File: src/test/compile-fail/return-from-diverging.rs (hunk 22 above)
Comment (canndrew) [inline src/test/compile-fail/return-from-diverging.rs]: I did delete some error messages in the process of implementing this RFC. There's no more "return in a function declared as diverging" error message for example.

File: src/test/compile-fail/call-fn-never-arg-wrong-type.rs (hunk 20 above)
Comment (canndrew) [inline src/test/compile-fail/call-fn-never-arg-wrong-type.rs]: I've added comments to all of them now. These tests are intended (for the most part) to follow the list you gave earlier.

File: src/test/run-pass/unit-fallback.rs [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Test that diverging types default to () (with feature(never_type) disabled).
```
Comment (nikomatsakis) [inline src/test/run-pass/unit-fallback.rs]: I still don't see where the diverging type variable arises in this test :)

**Comment by nikomatsakis** [comment]
@canndrew looking good -- main question is <https://github.com/rust-lang/rust/pull/35162/files#r74481879>.

File: src/test/run-pass/unit-fallback.rs (hunk 23 above)
Comment (canndrew) [inline src/test/run-pass/unit-fallback.rs]: Line 30. It's the type of `_`.

Comment (canndrew) [inline src/test/run-pass/unit-fallback.rs]: I've added a comment for it (and for the same thing in the never-fallback test).

**Comment by canndrew** [comment]
I think this is good to go then. For the tests that aren't marked off ..

> Dead-code warnings (are these implemented?) 

Not fully, `panic!()` et al. will trigger the warning because they return `!` directly though. This is already tested by preexisting tests.

> `match break { () => { } } // legal now, and afterwards I guess`

This already has a test: `run-pass/weird-exprs.rs`

>  Feature-gating tests:

These are the `never-fallback.rs` and `unit-fallback.rs` tests.

**Comment by nikomatsakis** [comment]
@canndrew r=me -- but needs rebase. Ping me or someone else to r+ when it's done :)

**Comment by arielb1** [comment]
```
running 2490 tests
Trace/breakpoint trap (core dumped)
make: *** [tmp/check-stage2-T-x86_64-unknown-linux-gnu-H-x86_64-unknown-linux-gnu-rpass.ok] Error 133
```

@bors retry
?

**Comment by eddyb** [comment]
@bors r-

@arielb1 This is most likely indicative of old trans being broken by this PR.

**Comment by arielb1** [comment]
Old trans apparently sometimes drops rvalue datums from other branches. You should use an lvalue datum instead.

fix:
```Rust
diff --git a/src/librustc_trans/expr.rs b/src/librustc_trans/expr.rs
index 0ea5715..18bafdb 100644
--- a/src/librustc_trans/expr.rs
+++ b/src/librustc_trans/expr.rs
@@ -385,7 +385,7 @@ fn apply_adjustments<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
             let mono_target = bcx.monomorphize(target);
             let llty = type_of::type_of(bcx.ccx(), mono_target);
             let dummy = C_undef(llty.ptr_to());
-            datum = Datum::new(dummy, mono_target, Rvalue::new(ByRef)).to_expr_datum();
+            datum = Datum::new(dummy, mono_target, Lvalue::new("never")).to_expr_datum();
         }
         AdjustReifyFnPointer => {
             match datum.ty.sty {
```

test case (sans assertions):
```Rust
pub struct Receiver(u32);
impl Drop for Receiver {
    fn drop(&mut self) {}
}

pub fn recv(f1: bool, f2: bool) -> Receiver {
    match f1 {
        false => Receiver(0),
        true => {
            match f2 {
                true => return Receiver(0),
                false => return Receiver(0),
            }
        }
    }
}

fn main() {}
```

**Comment by canndrew** [comment]
@eddyb Thanks!

**Comment by canndrew** [comment]
Also @arielb1 thanks!

**Comment by canndrew** [comment]
Anyone know what happened here? Has the problem with msvc wanting libm been fixed?

**Comment by retep998** [comment]
I really doubt msvc wants libm. If that legitimately happened I'd be quite worried. It is probably just some other bug causing a random "m" to appear in the command to the linker.

**Comment by durka** [comment]
For the record this indirectly introduced some issues (see https://github.com/rust-lang/rust/pull/35883#issuecomment-241473562) and parts of it will need to be semi-reverted (old `Diverging` branches replaced with new `.is_never()` branches).
